### PR TITLE
Jeremie/ri 484 hd dsfr structure du header

### PR DESCRIPTION
**Pull request by ledjay**
# Header + language modal refactor

## Redirections
- I had to rewrite some paths as /qui-sommes-nous becomes /mission-impact

## Header
- I cleaned the old menu structure, still some work to do to make it more readable tough ^^
- also had to make some css tricks to pass icons inside the menu links (we need a robust standalone Icon component plugged on the DSFR ones on by the way)

## New reusable components
- I made some new components for better code architecture, tried to focus on the single responsability pattern 
- I made a reusable LanguageSelector, it's DSRFized AND totally autonomous ^^
- I also had to work on a 100% custom dropdown component, I did my best to make it reusable and accessible, let me know if you're okay with the way it works : 

```
    <DropdownRoot key="language" onOpenChange={(open) => setLangMenuOpened(open)}>
      <DropdownTrigger asChild>
        <Button iconId="fr-icon-translate-2" priority="tertiary">
          {router.locale?.toLocaleUpperCase()}{" "}
          <i className={cls(langMenuOpened ? "fr-icon-arrow-up-s-line" : "fr-icon-arrow-down-s-line")} />
        </Button>
      </DropdownTrigger>
      <DropdownContent position="right">
        <LanguageSelector />
      </DropdownContent>
    </DropdownRoot>
```
## Language Modal
- I made a few tweaks to make sure it works with the new selector look

P.S: Sorry about the huge PR and the lack of atomic commits ^^'
 
 PPS: handles the jeremie/ri-487-hd-dsfr-structure-des-onglets ticket too
 
 PPPs : handles the jeremie/ri-488-hd-accessibilite-correctifs-accessibilite too

### CODE REVIEW

File: apps/client/redirects.js [hunk 1]
```diff
@@ -18,6 +18,7 @@ const urlTranslations = [
 ];
 
 const oldPathsRedirects = [
+  ["/fr/qui-sommes-nous", "/fr/mission-impact"],
```
Comment (matthieu-fesselier) [inline apps/client/redirects.js:21]: On doit peut être le faire en anglais aussi ?

File: apps/client/src/components/Backend/screens/routes.ts [hunk 2]
```diff
@@ -49,12 +49,24 @@ export const backendRoutes: BackendRouteType[] = [
     component: UserAdminStructure,
     restriction: [RoleName.ADMIN],
   },
+  {
```
Comment (matthieu-fesselier) [inline apps/client/src/components/Backend/screens/routes.ts:52]: Je ne comprends pas bien pourquoi cette partie est dupliquée ?

File: apps/client/src/components/Navigation/Navbar/QuickAccessMenu/LoginButton.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import { useSelector } from "react-redux";
+import { getPath } from "~/routes";
+import { userSelector } from "~/services/User/user.selectors";
+import styles from "./LoginButton.module.scss";
+
+const LoginButton = () => {
+  const { user } = useSelector(userSelector);
+  const { t } = useTranslation();
+  // eslint-disable-next-line no-console
+  console.log(user);
```
Comment (matthieu-fesselier) [inline apps/client/src/components/Navigation/Navbar/QuickAccessMenu/LoginButton.tsx]: on peut supprimer ça ?

File: apps/client/src/components/Navigation/Navbar/QuickAccessMenu/LoginButton.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  return (
+    <>
+      {user ? (
+        <Button
+          key="login"
+          priority="primary"
+          linkProps={{
+            // @ts-ignore
+            href: getPath("/backend/user-profile", router.locale), // TODO : fix the type route error here
```
Comment (matthieu-fesselier) [inline apps/client/src/components/Navigation/Navbar/QuickAccessMenu/LoginButton.tsx]: la fonction getPath me servait surtout à récupérer l'url traduite. Dans le cas du backend, comme il n'y a pas de trad, on peut mettre directement `/backend/user-profile/`, ce qui corrige l'erreur de type je pense

File: apps/client/src/components/UI/LanguageSelector/LanguageItem.tsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+interface LanguageItemProps {
+  item: GetLanguagesResponse;
+  className?: string;
+}
+
+const LanguageItem = memo(
+  forwardRef<HTMLButtonElement, LanguageItemProps>(({ item, className, ...props }, ref) => {
+    // eslint-disable-next-line no-console
+    // console.log(item);
```
Comment (matthieu-fesselier) [inline apps/client/src/components/UI/LanguageSelector/LanguageItem.tsx]: à supprimer ?

File: apps/client/src/components/UI/LanguageSelector/LanguageItem.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    const { t } = useTranslation();
+
+    const { changeLanguage } = useChangeLanguage();
+    const currentLanguage = router.locale || "fr";
+
+    const langues = useSelector(allLanguesSelector);
+    const notListenableLanguages = ["er"];
```
Comment (matthieu-fesselier) [inline apps/client/src/components/UI/LanguageSelector/LanguageItem.tsx:29]: on n'a pas cette langue, si ?

**Review by matthieu-fesselier — COMMENTED** [review]
Très cool ! J'ai mis quelques commentaires mais ce sont soit des questions, soit des petits console.log à retirer. 

Juste j'ai remarqué un petit bug de design en version mobile, sur les paddings horizontaux des premiers éléments : 
Capture d’écran 2024-12-11 à 09 55 43

tout bon sinon !

File: apps/client/src/components/UI/LanguageSelector/LanguageItem.tsx (hunk 6 above)
Comment (ledjay) [reply apps/client/src/components/UI/LanguageSelector/LanguageItem.tsx:29]: Si si, c'est le Tigrigna (Erythrée)

File: apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+const AccessibleNavigation = forwardRef<HTMLDivElement, AccessibleNavigationProps>(
+  (
+    { children, orientation = "vertical", className, onEscape, onNavigateOut, "aria-label": ariaLabel, ...props },
+    ref,
+  ) => {
+    const navRef = useRef<HTMLDivElement | null>(null);
+    const [activeIndex, setActiveIndex] = useState(0);
+
+    useImperativeHandle(ref, () => navRef.current as HTMLDivElement);
```
Comment (kaaloo) [inline apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx:41]: This usage of useImperativeHandle doesn't seem to acheive much.  Unless I'm missing something might as well use the ref directly.

File: apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+    const focusItem = useCallback((index: number) => {
+      const items = navRef.current?.querySelectorAll<HTMLElement>("[data-nav-item]");
```
Comment (kaaloo) [inline apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx:44]:
```suggestion
      const items = ref.current?.querySelectorAll<HTMLElement>("[data-nav-item]");
```

File: apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+      if (items && items[index]) {
+        items[index].focus();
+        setActiveIndex(index);
+      }
+    }, []);
+
+    const handleKeyDown = useCallback(
+      (e: KeyboardEvent<HTMLDivElement>) => {
+        const items = navRef.current?.querySelectorAll<HTMLElement>("[data-nav-item]");
```
Comment (kaaloo) [inline apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx:53]:
```suggestion
        const items = ref.current?.querySelectorAll<HTMLElement>("[data-nav-item]");
```

File: apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+};
+
+const AccessibleNavigationContext = createContext<AccessibleNavigationContextProps | undefined>(undefined);
+
+const AccessibleNavigation = forwardRef<HTMLDivElement, AccessibleNavigationProps>(
+  (
+    { children, orientation = "vertical", className, onEscape, onNavigateOut, "aria-label": ariaLabel, ...props },
+    ref,
+  ) => {
+    const navRef = useRef<HTMLDivElement | null>(null);
```
Comment (kaaloo) [inline apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx:37]:
```suggestion
```

**Review by kaaloo — APPROVED** [review]
Code review looks good, left some comments.  There is [a comment](https://github.com/refugies-info/karfur/pull/2502/files#r1879627481) from @matthieu-fesselier that hasn't been taken care of.

File: apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx (hunk 7 above)
Comment (ledjay) [reply apps/client/src/components/UI/AccessibleNavigation/AccessibleNavigation.tsx:41]: Thanks for your eagle-eye ^^
I just made a few tests to be sure.
If I don't use the imperativeHandle react loose the underlying div and the roving index mechanism won't work as expected

**Comment by ledjay** [comment]
@kaaloo 
Thanks for your feed-back, the useImperativeHandle seems overkill here indeed 😅

But I made a bunch of tests to make it work without useImperativeHandle unfortunately without it react looses track of the childrens nested inside the parent container and the RGAA's tabbing roving index don't work anymore.

It looks like the only way to keep this component structure and features but maybe there's another way ?

**Pull request merged**